contracts/DelegatedMigrator.sol

```diff
@@ -0,0 +1,134 @@
+pragma solidity ^0.5.16;
+import "./Owned.sol";
+import "./AddressResolver.sol";
+import "./Proxy.sol";
+import "./SystemStatus.sol";
+
+interface IMigration {
+    // pass the index to handle situations where the migration takes a few transactions to process
+    function migrate(uint256 index) external;
+
+    function numOfScripts() external view returns (uint);
+}
+
+contract DelegatedMigrator is Owned {
+    uint waitingPeriod;
+
+    struct Migration {
+        uint acceptedTimestamp;
+        IMigration target;
+        address dataStore;
+    }
+
+    AddressResolver resolver = AddressResolver(0);
+
+    mapping(bytes32 => Migration) proposals;
+
+    constructor(address _owner) public Owned(_owner) {
+        waitingPeriod = 3 hours;
+    }
+
+    function setWaitingPeriod(uint _waitingPeriod) external onlyOwner {
+        waitingPeriod = _waitingPeriod;
+    }
+
+    function propose(
+        bytes32 version,
+        IMigration target,
+        address dataStore
+    )
+        external
+    {
+        // Anyone can call
+        require(proposals[version].target == IMigration(0), "Cannot modify existing proposal");
+        require(address(target) != address(0), "Invalid target");
+        proposals[version] = Migration(
+                0,
+                target,
+                dataStore
+            );
+    }
+
+    function accept(bytes32 version) external onlyOwner {
+        require(proposals[version].target != IMigration(0), "invalid proposal");
+        proposals[version].acceptedTimestamp = now;
+    }
+
+    function reject(bytes32 version) external onlyOwner {
+        require(proposals[version].target != IMigration(0), "invalid proposal");
+        delete proposals[version];
+    }
+
+    function execute(bytes32 version, uint index) external onlyOwner {
+        Migration memory migration = proposals[version];
+        require(migration.acceptedTimestamp > 0, "Must be accepted to execute");
+        require(now > migration.acceptedTimestamp + waitingPeriod, "Waiting period not yet expired");
+
+        SystemStatus status = SystemStatus(resolver.getAddress("SystemStatus"));
+        if (index == 0) {
+          // deactivate system
+          status.suspendSystem(0);
+        }
+
+        (bool success, ) = address(migration.target).delegatecall(abi.encodePacked(migration.target.migrate.selector, version, index));
```

Comment:
IMO, using delegatecall here is dangerous.

This is done to, (as stated in a comment below) "make it easier to access storage". There is only 1 access to the migrator's storage for a simple read operation in the example migration. This _may_ be practical at some point, but unnecessary, since it opens up the possibility of a migration invading the migrator's state and potentially breaking it.

Reply:
UPDATE: As pointed out by Justin, having MigrationScript extend DelegateMigrator is also done so that the script is an owner as well, which actually makes calls like `proxy.setTarget(...)` possible.